### PR TITLE
fix: KeyError no longer occurs when using groupfolds for regression tasks.

### LINKED ISSUE #1254: Regression with Grouped kFold CV
Hello,

I have a _regression_ problem where I need to be able to use a grouped k-fold strategy (`sklearn.model_selection.GroupKFold`). However, I see in the docs that `split_type` for regression problems cannot be set to 'group' ?

Is there a way to run a regression problem with a grouped kFold approach in autoML?

Cheers!

### LINKED ISSUE #1384: [Bug]: KeyError raised when Using GroupKFold for Regression Tasks
### Describe the bug

Hi,

I've found an issue where group folds cannot be used for regression tasks. There are (as far as I know) four requirements for this error to occur.

1. The task is a regression task.
2. We have `split_type='group'`.
3. `y_train` (passed to `AutoML.fit()` is a Pandas Series.
4. The indices of `y_train` aren't equal to `range(len(y_train.shape[0]))` i.e. some indices may be out of order/missing.

Essentially when these criteria are met, we receive a KeyError, such as the one shown below:

Image

It's worth mentioning that if we don't use GroupKFolds (i.e. we set `split_type='uniform'`), then this issue doesn't occur - everything works well.

### Steps to reproduce

An example piece of code which generates this error is shown below. Essentially we use the Iris dataset as a regression dataset, and append a nonsensical cluster column to the dataset. We then split this data into a train and test set. Here we are using FLAML to perform the train and validation steps, whilst we withhold the test set to measure performance against afterwards.

```
from sklearn import datasets
import numpy as np
from flaml import AutoML
from sklearn.model_selection import train_test_split

iris_dict_data = datasets.load_iris(as_frame=True)  # numpy arrays
iris_data = iris_dict_data["frame"]  # pandas dataframe data + target
iris_data['cluster'] = np.random.randint(0, 5, iris_data.shape[0])
automl = AutoML()

X= iris_data[["sepal length (cm)","sepal width (cm)", "petal length (cm)"]].to_numpy()
y = iris_data["petal width (cm)"]
X_train, X_test, y_train, y_test, groups_train, groups_test = train_test_split(X, y, iris_data['cluster'], random_state=42)
automl_settings = {
    "max_iter":5,
    "time_budget":-1,
    "metric": 'r2',
    "task": 'regression',
    "log_file_name": "error.log",
    "log_type": "all",
    "estimator_list": ['lgbm'],
    "eval_method": "cv",
    "split_type":"group",
    "groups":groups_train
}

automl.fit(X_train, y_train, **automl_settings)
```

I've investigated this issue, and found the cause. Essentially when `y_train` is a Pandas series and the `split_type` is NOT `group`, the index of y_train is reset. Applying this same `reset_index()` when `split_type='group'` fixes the issue. I will create a PR for this, shortly

Again, it's worth highlighting that if we set `split_type='uniform'` above, then the code runs without any issues.

### Model Used

_No response_

### Expected Behavior

FLAML should be able to handle regression tasks with GroupKFold splits.

### Screenshots and logs

_No response_

### Additional Information

FLAML 2.3.2
Python 3.10.3

### PR DESCRIPTION
**Pull request by dannycg1996**
## Why are these changes needed?

Currently KeyErrors are raised (under certain conditions outlined in the linked issue) when using AutoML for regression tasks, with GroupKFold splits.
This was caused by the index not being reset for group folds.

As part of this PR, I've modified the if statement in `prepare_data` to ensure that we now also reset the index, regardless of the split type. This will ensure that this error doesn't occur when group folds are used, or when custom splitters are passed in. I've also added a test, which checks that my example code from the linked issue can now run without errors.

Hopefully that's merge-able - please let me know if you have any suggestions/comments.
Thanks!

## Related issue number

 Close #1384
Closes #1254
## Checks

- [X] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR (note the same in integrated in our CI checks).
- [X] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by thinkall — COMMENTED** [review]
Thank you @dannycg1996 for the PR. Please see my comments in #1384 .

**Review by thinkall — COMMENTED** [review]
Since we're allowing `group` for regression task. Do you mind update the related docstring as well? Thanks.

**Comment by dannycg1996** [comment]
Thanks @thinkall - I have updated the docs. Please let me know if I've missed anything!
Can I please ask what the plans are for future releases? We are building a project that depends on the latest FLAML changes.
Thanks again!

**Comment by thinkall** [comment]
> Thanks @thinkall - I have updated the docs. Please let me know if I've missed anything! Can I please ask what the plans are for future releases? We are building a project that depends on the latest FLAML changes. Thanks again!

I can make a release once your PR is merged.

**Review by thinkall — APPROVED** [review]

**Comment by dannycg1996** [comment]
That's great - thanks!
